### ISSUE TITLE
bug: CALLDATACOPY

### ISSUE THREAD

**Issue by greged93** [kkrt-labs/ef-tests#20]
# Bug Report

**Kakarot version:** [ee6458a](https://github.com/kkrt-labs/kakarot/commit/ee6458a7089f169334a7dba9665bcc5551507762)

**Current behavior:** 
Test for CALLDATACOPY provided by ethereum/tests fails during test `calldatacopy_d7g0v0_Shanghai` when transaction is expected to pass (see here, [post state](https://github.com/ethereum/tests/blob/develop/BlockchainTests/GeneralStateTests/VMTests/vmTests/calldatacopy.json#L8808) is different from [pre state](https://github.com/ethereum/tests/blob/develop/BlockchainTests/GeneralStateTests/VMTests/vmTests/calldatacopy.json#L8881)).

**Expected behavior:**
Test should pass.

**Steps to reproduce:**

- clone ef-tests: https://github.com/kkrt-labs/ef-tests
- generate katana dump of kakarot system and move the .katana folder to your ef-tests directory
- run 
```
make setup
make target=calldatacopy_d7g0v0_Shanghai ef-test
```
Failing output should be:
```
Transaction execution error: "Error in the called contract (0x04e2460d0debdacdde372cb598b4e4403939ca112c8720e1c2aca13f6ac09b9a):
Error at pc=0:7:
Got an exception while executing a hint.
Cairo traceback (most recent call last):
Unknown location (pc=0:163)
Unknown location (pc=0:149)

Error in the called contract (0x04e2460d0debdacdde372cb598b4e4403939ca112c8720e1c2aca13f6ac09b9a):
Error at pc=0:31:
Got an exception while executing a hint.
Cairo traceback (most recent call last):
Unknown location (pc=0:4606)
Unknown location (pc=0:4545)
Unknown location (pc=0:4374)
Unknown location (pc=0:376)

Error in the called contract (0x049c085e55e5ab05ea46e16706720852421e86c1f8aeeab2be257313c4e290f4):
Error at pc=0:2671:
Couldn't compute operand op1. Unknown value for memory cell 40:1
Cairo traceback (most recent call last):
Unknown location (pc=0:18120)
Unknown location (pc=0:18120)
Unknown location (pc=0:18120)
Unknown location (pc=0:18120)
Unknown location (pc=0:18120)
Unknown location (pc=0:18120)
Unknown location (pc=0:18120)
Unknown location (pc=0:18120)
Unknown location (pc=0:18120)
Unknown location (pc=0:18120)
Unknown location (pc=0:18120)
Unknown location (pc=0:18120)
Unknown location (pc=0:18120)
Unknown location (pc=0:18120)
Unknown location (pc=0:18120)
Unknown location (pc=0:18120)
Unknown location (pc=0:18120)
Unknown location (pc=0:18091)
Unknown location (pc=0:16584)
Unknown location (pc=0:3321)
"
[!] Case ~/code/rust/ef-tests/crates/ef-testing/ethereum-tests/BlockchainTests/GeneralStateTests/VMTests/vmTests/calldatacopy.json failed (description: calldatacopy): Test failed: failed test calldatacopy_d7g0v0_Shanghai: expected storage value 0x0000000000000000000000000000000000000000000000000000000badc0ffee, got 0x0000000000000000000000000000000000000000000000000000000000000000
```

**Comment by Eikix** [comment]
Tried to run this particular test in ef-tests, and kakarot main repo, it passes.
A new test that does not pass is: `calldatacopy_d4g0v0_Shanghai`.

Due to:
```
calldatacopy_d4g0v0_Shanghai balance mismatch for 0xa94f5374fce5edbc8e2a8697c15331677e6ebf0b: expected 0x0000000000000000000000000000000000000000000000000ba1a9ce0b9d0099, got 0x0000000000000000000000000000000000000000000000000ba1a9ce0b9d009a
test blockchain_tests::vm_tests ... FAILED
```

**Comment by Eikix** [comment]
Decoded trace:


/Users/eliastazartes/code/theodo/kakarot/.venv/lib/python3.9/site-packages/starkware/cairo/common/hash.cairo:16:18: Error at pc=0:7:
    let result = hash_ptr.result;
                 ^*************^
Got an exception while executing a hint: Hint Error: Error in the called contract (0x01ee6133613c44c36bd4c2808e9ac2079b18cc9e6dd54d54c7e684175b32fe8b):
Error at pc=0:31:
Got an exception while executing a hint: Hint Error: Error in the called contract (0x0000000000000000000000000000000000000000000000000000000000000001):
/Users/eliastazartes/code/theodo/kakarot/src/utils/utils.cairo:270:34: Error at pc=0:2786:
        memcpy(dst=new_data, src=data + data_offset, len=len);
                                 ^****************^
Operation failed: 41:0 + 340282366920938463463374607431768211450, maximum offset value exceeded
Cairo traceback (most recent call last):
/Users/eliastazartes/code/theodo/kakarot/src/kakarot/evm.cairo:626:20
            return run(ctx=ctx);
                   ^**********^
/Users/eliastazartes/code/theodo/kakarot/src/kakarot/evm.cairo:626:20
            return run(ctx=ctx);
                   ^**********^
/Users/eliastazartes/code/theodo/kakarot/src/kakarot/evm.cairo:626:20
            return run(ctx=ctx);
                   ^**********^
/Users/eliastazartes/code/theodo/kakarot/src/kakarot/evm.cairo:626:20
            return run(ctx=ctx);
                   ^**********^
/Users/eliastazartes/code/theodo/kakarot/src/kakarot/evm.cairo:626:20
            return run(ctx=ctx);
                   ^**********^
/Users/eliastazartes/code/theodo/kakarot/src/kakarot/evm.cairo:626:20
            return run(ctx=ctx);
                   ^**********^
/Users/eliastazartes/code/theodo/kakarot/src/kakarot/evm.cairo:626:20
            return run(ctx=ctx);
                   ^**********^
/Users/eliastazartes/code/theodo/kakarot/src/kakarot/evm.cairo:626:20
            return run(ctx=ctx);
                   ^**********^
/Users/eliastazartes/code/theodo/kakarot/src/kakarot/evm.cairo:626:20
            return run(ctx=ctx);
                   ^**********^
/Users/eliastazartes/code/theodo/kakarot/src/kakarot/evm.cairo:626:20
            return run(ctx=ctx);
                   ^**********^
/Users/eliastazartes/code/theodo/kakarot/src/kakarot/evm.cairo:626:20
            return run(ctx=ctx);
                   ^**********^
/Users/eliastazartes/code/theodo/kakarot/src/kakarot/evm.cairo:626:20
            return run(ctx=ctx);
                   ^**********^
/Users/eliastazartes/code/theodo/kakarot/src/kakarot/evm.cairo:626:20
            return run(ctx=ctx);
                   ^**********^
/Users/eliastazartes/code/theodo/kakarot/src/kakarot/evm.cairo:626:20
            return run(ctx=ctx);
                   ^**********^
/Users/eliastazartes/code/theodo/kakarot/src/kakarot/evm.cairo:626:20
            return run(ctx=ctx);
                   ^**********^
/Users/eliastazartes/code/theodo/kakarot/src/kakarot/evm.cairo:626:20
            return run(ctx=ctx);
                   ^**********^
/Users/eliastazartes/code/theodo/kakarot/src/kakarot/evm.cairo:626:20
            return run(ctx=ctx);
                   ^**********^
/Users/eliastazartes/code/theodo/kakarot/src/kakarot/evm.cairo:623:44
        let ctx: model.ExecutionContext* = decode_and_execute(ctx=ctx);
                                           ^*************************^
/Users/eliastazartes/code/theodo/kakarot/src/kakarot/evm.cairo:208:9
        call EnvironmentalInformation.exec_calldatacopy;  // 0x37
        ^*********************************************^
/Users/eliastazartes/code/theodo/kakarot/src/kakarot/instructions/environmental_information.cairo:295:38
        let sliced_calldata: felt* = Helpers.slice_data(
                                     ^*****************^

Cairo traceback (most recent call last):
/Users/eliastazartes/code/theodo/kakarot/src/kakarot/execution_context.cairo:394:29
            program_counter=self.program_counter,
                            ^******************^
/Users/eliastazartes/code/theodo/kakarot/src/kakarot/execution_context.cairo:304:20
        let self = ExecutionContext.increment_program_counter(self=self, inc_value=len);
                   ^******************************************************************^
Unknown location (pc=0:4428)
/Users/eliastazartes/code/theodo/kakarot/.venv/lib/python3.9/site-packages/starkware/cairo/common/math_cmp.cairo:54:9
        ap += 25;
        ^******^

Error in the called contract (0x0000000000000000000000000000000000000000000000000000000000000001):
/Users/eliastazartes/code/theodo/kakarot/src/utils/utils.cairo:270:34: Error at pc=0:2786:
        memcpy(dst=new_data, src=data + data_offset, len=len);
                                 ^****************^
Operation failed: 41:0 + 340282366920938463463374607431768211450, maximum offset value exceeded
Cairo traceback (most recent call last):
/Users/eliastazartes/code/theodo/kakarot/src/kakarot/evm.cairo:626:20
            return run(ctx=ctx);
                   ^**********^
/Users/eliastazartes/code/theodo/kakarot/src/kakarot/evm.cairo:626:20
            return run(ctx=ctx);
                   ^**********^
/Users/eliastazartes/code/theodo/kakarot/src/kakarot/evm.cairo:626:20
            return run(ctx=ctx);
                   ^**********^
/Users/eliastazartes/code/theodo/kakarot/src/kakarot/evm.cairo:626:20
            return run(ctx=ctx);
                   ^**********^
/Users/eliastazartes/code/theodo/kakarot/src/kakarot/evm.cairo:626:20
            return run(ctx=ctx);
                   ^**********^
/Users/eliastazartes/code/theodo/kakarot/src/kakarot/evm.cairo:626:20
            return run(ctx=ctx);
                   ^**********^
/Users/eliastazartes/code/theodo/kakarot/src/kakarot/evm.cairo:626:20
            return run(ctx=ctx);
                   ^**********^
/Users/eliastazartes/code/theodo/kakarot/src/kakarot/evm.cairo:626:20
            return run(ctx=ctx);
                   ^**********^
/Users/eliastazartes/code/theodo/kakarot/src/kakarot/evm.cairo:626:20
            return run(ctx=ctx);
                   ^**********^
/Users/eliastazartes/code/theodo/kakarot/src/kakarot/evm.cairo:626:20
            return run(ctx=ctx);
                   ^**********^
/Users/eliastazartes/code/theodo/kakarot/src/kakarot/evm.cairo:626:20
            return run(ctx=ctx);
                   ^**********^
/Users/eliastazartes/code/theodo/kakarot/src/kakarot/evm.cairo:626:20
            return run(ctx=ctx);
                   ^**********^
/Users/eliastazartes/code/theodo/kakarot/src/kakarot/evm.cairo:626:20
            return run(ctx=ctx);
                   ^**********^
/Users/eliastazartes/code/theodo/kakarot/src/kakarot/evm.cairo:626:20
            return run(ctx=ctx);
                   ^**********^
/Users/eliastazartes/code/theodo/kakarot/src/kakarot/evm.cairo:626:20
            return run(ctx=ctx);
                   ^**********^
/Users/eliastazartes/code/theodo/kakarot/src/kakarot/evm.cairo:626:20
            return run(ctx=ctx);
                   ^**********^
/Users/eliastazartes/code/theodo/kakarot/src/kakarot/evm.cairo:626:20
            return run(ctx=ctx);
                   ^**********^
/Users/eliastazartes/code/theodo/kakarot/src/kakarot/evm.cairo:623:44
        let ctx: model.ExecutionContext* = decode_and_execute(ctx=ctx);
                                           ^*************************^
/Users/eliastazartes/code/theodo/kakarot/src/kakarot/evm.cairo:208:9
        call EnvironmentalInformation.exec_calldatacopy;  // 0x37
        ^*********************************************^
/Users/eliastazartes/code/theodo/kakarot/src/kakarot/instructions/environmental_information.cairo:295:38
        let sliced_calldata: felt* = Helpers.slice_data(
                                     ^*****************^

Cairo traceback (most recent call last):
/Users/eliastazartes/code/theodo/kakarot/.venv/lib/python3.9/site-packages/starkware/cairo/common/math.cairo:130:15
    let low = [range_check_ptr];
              ^***************^
Unknown location (pc=0:149)

Error in the called contract (0x01ee6133613c44c36bd4c2808e9ac2079b18cc9e6dd54d54c7e684175b32fe8b):
Error at pc=0:31:
Got an exception while executing a hint: Hint Error: Error in the called contract (0x0000000000000000000000000000000000000000000000000000000000000001):
/Users/eliastazartes/code/theodo/kakarot/src/utils/utils.cairo:270:34: Error at pc=0:2786:
        memcpy(dst=new_data, src=data + data_offset, len=len);
                                 ^****************^
Operation failed: 41:0 + 340282366920938463463374607431768211450, maximum offset value exceeded
Cairo traceback (most recent call last):
/Users/eliastazartes/code/theodo/kakarot/src/kakarot/evm.cairo:626:20
            return run(ctx=ctx);
                   ^**********^
/Users/eliastazartes/code/theodo/kakarot/src/kakarot/evm.cairo:626:20
            return run(ctx=ctx);
                   ^**********^
/Users/eliastazartes/code/theodo/kakarot/src/kakarot/evm.cairo:626:20
            return run(ctx=ctx);
                   ^**********^
/Users/eliastazartes/code/theodo/kakarot/src/kakarot/evm.cairo:626:20
            return run(ctx=ctx);
                   ^**********^
/Users/eliastazartes/code/theodo/kakarot/src/kakarot/evm.cairo:626:20
            return run(ctx=ctx);
                   ^**********^
/Users/eliastazartes/code/theodo/kakarot/src/kakarot/evm.cairo:626:20
            return run(ctx=ctx);
                   ^**********^
/Users/eliastazartes/code/theodo/kakarot/src/kakarot/evm.cairo:626:20
            return run(ctx=ctx);
                   ^**********^
/Users/eliastazartes/code/theodo/kakarot/src/kakarot/evm.cairo:626:20
            return run(ctx=ctx);
                   ^**********^
/Users/eliastazartes/code/theodo/kakarot/src/kakarot/evm.cairo:626:20
            return run(ctx=ctx);
                   ^**********^
/Users/eliastazartes/code/theodo/kakarot/src/kakarot/evm.cairo:626:20
            return run(ctx=ctx);
                   ^**********^
/Users/eliastazartes/code/theodo/kakarot/src/kakarot/evm.cairo:626:20
            return run(ctx=ctx);
                   ^**********^
/Users/eliastazartes/code/theodo/kakarot/src/kakarot/evm.cairo:626:20
            return run(ctx=ctx);
                   ^**********^
/Users/eliastazartes/code/theodo/kakarot/src/kakarot/evm.cairo:626:20
            return run(ctx=ctx);
                   ^**********^
/Users/eliastazartes/code/theodo/kakarot/src/kakarot/evm.cairo:626:20
            return run(ctx=ctx);
                   ^**********^
/Users/eliastazartes/code/theodo/kakarot/src/kakarot/evm.cairo:626:20
            return run(ctx=ctx);
                   ^**********^
/Users/eliastazartes/code/theodo/kakarot/src/kakarot/evm.cairo:626:20
            return run(ctx=ctx);
                   ^**********^
/Users/eliastazartes/code/theodo/kakarot/src/kakarot/evm.cairo:626:20
            return run(ctx=ctx);
                   ^**********^
/Users/eliastazartes/code/theodo/kakarot/src/kakarot/evm.cairo:623:44
        let ctx: model.ExecutionContext* = decode_and_execute(ctx=ctx);
                                           ^*************************^
/Users/eliastazartes/code/theodo/kakarot/src/kakarot/evm.cairo:208:9
        call EnvironmentalInformation.exec_calldatacopy;  // 0x37
        ^*********************************************^
/Users/eliastazartes/code/theodo/kakarot/src/kakarot/instructions/environmental_information.cairo:295:38
        let sliced_calldata: felt* = Helpers.slice_data(
                                     ^*****************^

Cairo traceback (most recent call last):
/Users/eliastazartes/code/theodo/kakarot/src/kakarot/execution_context.cairo:394:29
            program_counter=self.program_counter,
                            ^******************^
/Users/eliastazartes/code/theodo/kakarot/src/kakarot/execution_context.cairo:304:20
        let self = ExecutionContext.increment_program_counter(self=self, inc_value=len);
                   ^******************************************************************^
Unknown location (pc=0:4428)
/Users/eliastazartes/code/theodo/kakarot/.venv/lib/python3.9/site-packages/starkware/cairo/common/math_cmp.cairo:54:9
        ap += 25;
        ^******^

Error in the called contract (0x0000000000000000000000000000000000000000000000000000000000000001):
/Users/eliastazartes/code/theodo/kakarot/src/utils/utils.cairo:270:34: Error at pc=0:2786:
        memcpy(dst=new_data, src=data + data_offset, len=len);
                                 ^****************^
Operation failed: 41:0 + 340282366920938463463374607431768211450, maximum offset value exceeded
Cairo traceback (most recent call last):
/Users/eliastazartes/code/theodo/kakarot/src/kakarot/evm.cairo:626:20
            return run(ctx=ctx);
                   ^**********^
/Users/eliastazartes/code/theodo/kakarot/src/kakarot/evm.cairo:626:20
            return run(ctx=ctx);
                   ^**********^
/Users/eliastazartes/code/theodo/kakarot/src/kakarot/evm.cairo:626:20
            return run(ctx=ctx);
                   ^**********^
/Users/eliastazartes/code/theodo/kakarot/src/kakarot/evm.cairo:626:20
            return run(ctx=ctx);
                   ^**********^
/Users/eliastazartes/code/theodo/kakarot/src/kakarot/evm.cairo:626:20
            return run(ctx=ctx);
                   ^**********^
/Users/eliastazartes/code/theodo/kakarot/src/kakarot/evm.cairo:626:20
            return run(ctx=ctx);
                   ^**********^
/Users/eliastazartes/code/theodo/kakarot/src/kakarot/evm.cairo:626:20
            return run(ctx=ctx);
                   ^**********^
/Users/eliastazartes/code/theodo/kakarot/src/kakarot/evm.cairo:626:20
            return run(ctx=ctx);
                   ^**********^
/Users/eliastazartes/code/theodo/kakarot/src/kakarot/evm.cairo:626:20
            return run(ctx=ctx);
                   ^**********^
/Users/eliastazartes/code/theodo/kakarot/src/kakarot/evm.cairo:626:20
            return run(ctx=ctx);
                   ^**********^
/Users/eliastazartes/code/theodo/kakarot/src/kakarot/evm.cairo:626:20
            return run(ctx=ctx);
                   ^**********^
/Users/eliastazartes/code/theodo/kakarot/src/kakarot/evm.cairo:626:20
            return run(ctx=ctx);
                   ^**********^
/Users/eliastazartes/code/theodo/kakarot/src/kakarot/evm.cairo:626:20
            return run(ctx=ctx);
                   ^**********^
/Users/eliastazartes/code/theodo/kakarot/src/kakarot/evm.cairo:626:20
            return run(ctx=ctx);
                   ^**********^
/Users/eliastazartes/code/theodo/kakarot/src/kakarot/evm.cairo:626:20
            return run(ctx=ctx);
                   ^**********^
/Users/eliastazartes/code/theodo/kakarot/src/kakarot/evm.cairo:626:20
            return run(ctx=ctx);
                   ^**********^
/Users/eliastazartes/code/theodo/kakarot/src/kakarot/evm.cairo:626:20
            return run(ctx=ctx);
                   ^**********^
/Users/eliastazartes/code/theodo/kakarot/src/kakarot/evm.cairo:623:44
        let ctx: model.ExecutionContext* = decode_and_execute(ctx=ctx);
                                           ^*************************^
/Users/eliastazartes/code/theodo/kakarot/src/kakarot/evm.cairo:208:9
        call EnvironmentalInformation.exec_calldatacopy;  // 0x37
        ^*********************************************^
/Users/eliastazartes/code/theodo/kakarot/src/kakarot/instructions/environmental_information.cairo:295:38
        let sliced_calldata: felt* = Helpers.slice_data(
                                     ^*****************^

**Comment by Eikix** [comment]
TL;DR
```

/Users/eliastazartes/code/theodo/kakarot/src/kakarot/evm.cairo:208:9
        call EnvironmentalInformation.exec_calldatacopy;  // 0x37
        ^*********************************************^
/Users/eliastazartes/code/theodo/kakarot/src/kakarot/instructions/environmental_information.cairo:295:38
        let sliced_calldata: felt* = Helpers.slice_data(
                                     ^*****************^

# ====>>> slice_data:
memcpy(dst=new_data, src=data + data_offset, len=len);
                                 ^****************^
Operation failed: 41:0 + 340282366920938463463374607431768211450, maximum offset value exceeded

```

**Comment by greged93** [comment]
linked to #554
